### ISSUE TITLE
[sklearn#34125] ENH: L1 ball  Centroids Projection in MiniBatchKMeans (Sculley, 2010)

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#351]
**Mirrored from [scikit-learn#34125](https://github.com/scikit-learn/scikit-learn/issues/34125)**

Original title: ENH: L1 ball  Centroids Projection in MiniBatchKMeans (Sculley, 2010)
Original labels: New Feature, Needs Decision, module:cluster
Created: 2026-05-26T13:41:44Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the workflow you want to enable

I want to cluster high-dimensional sparse data to return **sparse centroids**. Typically a TF-IDF transform over a vocab of 10K to 100K.  

Today `MiniBatchKMeans` returns dense centroids with low nnz even when the data is 99% sparse.  

#### Benefits:
1. **Memory.** with `n_clusters=10_000` and `n_features=100_000` that cluster centers array would be 80GB alone which with 90% sparsity would become 8GB. It is what the Sculley paper designed `MiniBatchKMeans` to handle.  
2. **Fit cost.** Projection adds <5% to fit time. 95% sparse centroids means 95% fewer multiply-adds per assignment dot product. Turning that into wall-time requires the assignment loop to skip zero centroid entries (same pattern as `lloyd_iter_chunked_sparse`), which is a follow-on in the PR.

3. **Interpretability.** A sparse centroids with 50 nnz is intuitively easier to understand than one with `10_000`.

```python
mbk = MiniBatchKMeans(l1_radius=1).fit(X_tfidf)
mbk.cluster_centers_ # dense ndarray, but 90% are exactly 0.0

```
`l1_radius` being the radius of L1 ball that after each iteration the centroids gets projected on to.  
The idea is from Sculley 2010, where MiniBatchKMeans itself comes from.  
This would cause the centroids to be sparse which is useful for scaling and a natural fit for already sparse data(e.g. TF-IDF transformed text).  
Besides that it can also act as mild regularization.

Sketch of expected empirical impact (preliminary; concrete benchmark belongs in the PR):     
 - 20 Newsgroups TF-IDF, k=20, p~=24,000: at l1_radius=3.0, centroid sparsity rises from ~0%   
 to ~87% with `adjusted_mutual_info_score` ~1% better than plain MiniBatchKMeans.
 - RCV1 train split, k=4, p~=47,000: centroid sparsity ~96% at l1_radius=3.0, negligible       
 AMI/ARI delta.

### Describe your proposed solution

Add one optional parameter, `l1_radius` `float` or `None`, that defaults to `None` to `MiniBatchKMeans`.  
When `None` the estimator behaves the same way as today.  
When set to `r` every centroid update is followed by a Euclidean projection of each centroid onto L1 ball ${x: norm_1(x) <= r}$.  
This is from Section 5 of Sculley(2010), cited as the basis of `MiniBatchKMeans` itself in sklearn docs.  

#### Where would the change go:
After the EMA update inside `_minibatch_update_dense` / `_minibatch_update_sparse`, apply the projection to each updated row of `centers_new` before  sklearn's convergence diff is computed.                                           

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [67](https://github.com/godofecht/flow-scikit/issues/67) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
